---
title: "scPagwas"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output:
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# scPagwas

**scPagwas**, a polygenic principal component-scoring method that links scRNA-seq data with large-scale GWAS summary statistics to prioritize genetics-modulated cells associated with complex diseases.
**scPagwas** is able to prioritize disease-associated individual cells by integrating the scRNA-seq data with polygenic signals from GWAS.


![Graphical abstract](./docs/reference/figures/workflow_20220222.png)

## Installation

You can install the released version of scPagwas from [github](https://github.com/dengchunyu/scPagwas) with:

``` r
devtools::install_github("dengchunyu/scPagwas")

```

## Example 

### 1.Preprogress and create internal storage

```{r message=FALSE,results = "hide",warning = FALSE}
 library(scPagwas)
 library(ggplot2)
 suppressMessages(library(Seurat))
 suppressMessages(library("dplyr"))
 #Input pathway gene list, you can construct with youself.
 data(Genes_by_pathway_kegg)
 #gene annotation files.
 data(block_annotation)
 #LD data
 data(chrom_ld)
#
 #1.start to run the wrapper functions for preprogress.
 Pagwas_data<-scPagwas_main(Pagwas = NULL,
                     gwas_data =system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"),
                     Single_data =system.file("extdata", "scRNAexample.rds", package = "scPagwas"),
                     output.prefix="test",
                     output.dirs="scPagwastest_output",
                     block_annotation = block_annotation,
                     assay="RNA",
                     Pathway_list=Genes_by_pathway_kegg,
                     chrom_ld = chrom_ld,
                     singlecell=T,
                     seruat_return=T,
                     celltype=T,
                     ncores = 2,
                     split_n=1)

```
 
#### Visualize the celltypes results.

 1.barplot
```{r Bootstrap_P_Barplot, message=FALSE,warning = FALSE}

Bootstrap_P_Barplot(p_results=Pagwas_data@misc$bootstrap_results$bp_value[-1],
                    p_names=rownames(Pagwas_data@misc$bootstrap_results)[-1],
                    figurenames = NULL,
                    width = 5,
                    height = 7,
                    do_plot=T,
                    title = "Test scPagwas")

```


#### Visualize the scPagwas_main results.
##### Visualize the scPagwas_score of single cell data in UMAP or TSNE plot. 

```{r scPagwas_Visualization,message=FALSE,warning = FALSE}
 require("RColorBrewer")
 require("Seurat")
 require("SeuratObject")
 require("ggsci")
 #check the objects

Pagwas_data <- ScaleData(Pagwas_data,assay ="RNA")
Pagwas_data<-FindVariableFeatures(Pagwas_data,nfeatures = 2000,assay ="RNA")

Pagwas_data <- suppressMessages(Seurat::RunPCA(object = Pagwas_data, assay =  "RNA", npcs = 50))

Pagwas_data <- suppressMessages(Seurat::RunUMAP(object = Pagwas_data,  assay =  "RNA", reduction = "pca", dims = 1:50))


 DimPlot(Pagwas_data,group.by = "anno",pt.size=1,reduction="umap",label = T, repel=TRUE)+ 
 umap_theme()+ggtitle("Test")+labs(x="TSNE",y="")+theme(aspect.ratio=1)
 
 
 scPagwas_Visualization(Single_data=Single_data,
                                   p_thre = 0.05,
                                   FigureType = "umap",
                                   width = 7,
                                   height = 7,
                                   lowColor = "white", 
                        highColor = "red",
                        output.dirs="monosimu",
                                   size = 0.5,
                                   do_plot = T)
 
```

##### Plot the barplot of the proportion of positive Cells in celltypes

```{r bar_positie_nagtive,message=FALSE,warning = FALSE,fig.height=6, fig.width=6}
library("RColorBrewer")
library("ggplot2")

scRNAexample$CellsrankPvalue<-Pagwas$CellsrankPvalue[colnames(scRNAexample)]
scRNAexample$positiveCells<-rep(0,ncol(scRNAexample))
scRNAexample$positiveCells[scRNAexample$CellsrankPvalue <= 0.01]<-1

plot_bar_positie_nagtive(seurat_obj=scRNAexample,
                              var_ident="positiveCells",
                              var_group="anno",
                              vec_group_colors=NULL,
                              f_color=colorRampPalette(brewer.pal(n=10, name="RdYlBu")),
                              do_plot = T)
```

##### Plot the barplot of the proportion of celltypes in positive Cell 

```{r bar_positie_nagtive2,message=FALSE,warning = FALSE}
plot_bar_positie_nagtive(seurat_obj=scRNAexample,
                              var_ident="anno",
                              var_group="positiveCells",
                              vec_group_colors=c("#E8D0B3","#7EB5A6"),
                              do_plot = T)
```


##### Plot the top5 heritability correlation genes in celltypes

```{r vln_Corgenes,message=FALSE,fig.height=6, fig.width=8,warning = FALSE}
top5genes<-rownames(Pagwas$gene_heritability_correlation)[order(Pagwas$gene_heritability_correlation,decreasing = T)[1:5]]
plot_vln_Corgenes(seurat_obj=scRNAexample,
             assay="RNA", slot="data",
             var_group="anno",
             vec_features=top5genes,
             vec_group_colors= pal_d3(alpha =0.5)(10),
             do_plot = T
             )

```

