---
title: "The test for prune ad example and visulization"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output:
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}

---

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
 collapse = TRUE,
 comment = "#>",
  fig.path = "figures/vignette-",
 out.width = "60%"
)

```

## the process for some examples, not run and for Record
```{r setup, eval = FALSE}
###############not run
#' scRNAexample
#'
#' @description
#' A single cell example data from \code{
#' liver_data<-readRDS("/share/pub/qiuf/COVID/liver/Rpoly.rds")
#' #set Idents
#' liver_data@meta.data$annotation <- liver_data@meta.data$anno
#' Idents(liver_data) <- liver_data@meta.data$anno
#' #normalization
#' liver_data <- NormalizeData(liver_data, normalization.method = "LogNormalize", scale.factor = 10000)
#' #correct the cell names, need no specific signal
#' anno <- liver_data@meta.data$annotation
#'  anno<-str_replace_all(anno,"-",".")
#'  anno<-str_replace_all(anno," ",".")
#' anno[which(anno=="Inflammatory.monocyte/macrophages")]<-"Inflammatory.monocyte.macrophages"
#' Celltype_anno$annotation<-str_replace_all(anno,"\\+",".")
#' Idents(liver_data) <- as.factor(anno)
#' liver_data<-subset(liver_data,idents = c("Stellate.cell","unknow","γδ.T.cell"), invert = TRUE)
#' scRNAexample<-liver_data[,sample(ncol(liver_data),1000)]
#'
#'  for (x in 1:length(unique(scRNAexample@meta.data$anno))){
#'  newnames <- paste0("Celltype",x)
#'  oldnames<-unique(scRNAexample@meta.data$anno)[x]
#'  scRNAexample@meta.data$anno[which(scRNAexample@meta.data$anno==oldnames)]<-newnames
#'  }
#'  Idents(scRNAexample)<-scRNAexample@meta.data$anno
#'  scRNAexample@meta.data<-scRNAexample@meta.data[,1:15]
#'
#'  save(scRNAexample,file = "E:/RPakage/scPagwas/data/scRNAexample.RData")
#'
#' }
#'
#' @docType data
#' @name scRNAexample
#' @usage data(scRNAexample)
#' @format a Seruat.
#' @source Generated from PBC
#' @keywords datasets
#' @examples
#' data(scRNAexample)
#' str(scRNAexample)
NULL

#' GWAS_summ_example
#'
#' @description
#' A GWAS summary data from \code{
#' library(readr)
#' library(dplyr)
#' GWAS_liver<-read_delim("/share/pub/mayl/Sherlock/XiangbingYu_Primary_biliary_cholangitis/PBC_GWAS_UKBiobank_summary_final",delim="\t")
#' #select the specific coloumn names: chrom，pos，rsid，beta，se，maf
#' GWAS_liver<-GWAS_liver[,c("CHR","POS","SNP","p","SE")]
#' colnames(GWAS_liver)<-c("chrom","pos","rsid","p","se")
#' COVOD19<-read_delim("/share/pub/mayl/01_COVID19_GWAS_round4/COVID19_HGI_B2_ALL_leave_23andme_20201020.b37.txt",delim="\t")
#' COVOD19<-COVOD19[,c("all_meta_AF","rsid")]
#' colnames(COVOD19)<-c("maf","rsid")
#' ##get beta
#' GWAS_liver$beta<-GWAS_liver$se*qchisq(GWAS_liver$p,1,lower.tail=F)
#' GWAS_liver_inner <- inner_join(GWAS_liver,COVOD19,by="rsid")
#' GWAS_liver_inner <- as.data.frame(GWAS_liver_inner)
#' GWAS_summ_example <- as.data.frame(GWAS_liver_inner[sample(nrow(GWAS_liver_inner),10000),])
#' save(GWAS_summ_example,file="/share/pub/dengcy/GWAS_Multiomics/pagwas/data/GWAS_summ_example.RData")
#' }
#'
#' @docType data
#' @name GWAS_summ_example
#' @usage data(GWAS_summ_example)
#' @format a data.frame.
#' @source Generated from PBC_GWAS_UKBiobank_summary_final
#' @keywords datasets
#' @examples
#' data(GWAS_summ_example)
#' str(GWAS_summ_example)
NULL


```


# 1. Data progress（not run）
导入单细胞数据并对数据进行预处理，要求为标准化后的数据，assays 包含RNA或者SCT，Idents必须为细胞类型的注释。
为了后续可视化方便需要对单细胞进行降维处理，包括TSNE和UMAP降维方式。
```{r realexample, eval=FALSE}
 library(scPagwas)
 library(Seurat)
 library(parallel)
#library(SeuratData)
data(Genes_by_pathway_kegg)
data(block_annotation)
data(chrom_ld)
Single_data =readRDS("E:/RPakage/scPagwas/inst/extdata/GSE138852_ad.rds")
Single_data <- FindVariableFeatures(Single_data,nfeatures = 3000)
Single_data <- NormalizeData(Single_data, normalization.method = "LogNormalize", scale.factor = 10000)
Single_data <- ScaleData(Single_data)
Single_data <- RunPCA(object = Single_data, assay = "RNA", npcs = 50)
Single_data <- RunTSNE(object = Single_data,assay =  "RNA", reduction = "pca",dims = 1:50)
Single_data <- RunUMAP(object = Single_data, assay = "RNA", reduction = "pca",dims = 1:50)
save(Single_data,file = "E:/RPakage/scPagwas/inst/extdata/GSE138852_ad.rds")

Single_data<-subset(Single_data,idents = c("mg","OPC","astro"))

```


# 1.Main functions v1.7.3
```{r eval=FALSE}
 ##############真实例子：
Pagwas_seu<-scPagwas_main(Pagwas = NULL,
                    gwas_data ="E:/RPakage/scPagwas/inst/extdata/AD_prune_gwas_data.txt",
                   output.prefix="adPrunetest",
                   output.dirs="adPrunetest",
                   block_annotation = block_annotation,
                   assay="RNA",
                   Single_data ="E:/RPakage/scPagwas/inst/extdata/GSE138852_ad.rds",
                   ncores=5,
                   Pathway_list=Genes_by_pathway_kegg,
                  chrom_ld = chrom_ld,
                  singlecell=T,
                  celltype=T)

save(Pagwas_seu,file="E:/RPakage/scPagwas/data/Pagwas_AD_prune.RData")
```


# Visualize the result
## Visualize the scPagwas_score for single cell
首先产生单细胞单纯映射细胞类型位置和颜色
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
require("RColorBrewer")
require("SeuratObject")
require("ggsci")
library("RColorBrewer")
library("ggplot2")
#load("E:/RPakage/scPagwas/data/Pagwas_AD_prune.RData")

DimPlot(Pagwas_seu,reduction="umap",group.by = "oupSample.cellType",pt.size=0.8,
        label = TRUE, repel=TRUE)+ labs(x="TSNE",y="")+
        scale_color_d3() +
        theme(aspect.ratio=1)

```

之后产生两者单细胞降维可视化图，第一张为scPagwas_score得分映射，第二章为阳性细胞映射
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}

#setwd("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test")
 scPagwas_Visualization(Single_data=Pagwas,
                                   p_thre = 0.05,
                                   FigureType = "tsne",
                                   width = 7,
                                   height = 7,
                                   lowColor = "#000957", 
                        highColor = "#EBE645",
                        output.dirs="adPrunetest",
                                   size = 0.5,
                                   do_plot = T)
```

## positive cell percent

### 阳性细胞和阴性细胞中不同细胞类型的比例
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}

plot_bar_positie_nagtive(seurat_obj=Pagwas,
                              var_ident="positiveCells",
                              var_group="oupSample.cellType",
                              vec_group_colors=NULL,
                              f_color=colorRampPalette(brewer.pal(n=10, name="RdYlBu")),
                              do_plot = T,
                         output.prefix="1",
                          output.dirs = "adPrunetest")

plot_bar_positie_nagtive(seurat_obj=Pagwas,
                              var_ident="oupSample.cellType",
                              var_group="positiveCells",
                              vec_group_colors=c("#E8D0B3","#7EB5A6"),
                         output.prefix="2",
                         output.dirs = "adPrunetest",
                              do_plot = T)
```


### 可视化通路降维数据中不同细胞类型的分布

```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
DimPlot(Pagwas,group.by = "oupSample.cellType",label = F)+ 
        theme(aspect.ratio=1)


```

## 通路数据降维可视化三种结果的映射
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
Pagwas<-preprogress_PaSingle_data(Single_data=Pagwas,
                       npcs=10,
                       nfeatures=20)
scPagwasPaHeritability_Visualization(Single_data=Pagwas,
                                    p_thre = 0.05,
                                    output.dirs="adPrunetest",
                                    FigureType = "umap",
                                    width = 7,
                                    height = 7,
                                    lowColor="#8479E1",
                                    midColor="#F7F5F2",
                                    highColor="#FD5D5D",
                                    p_color= "#E45826",
                                    size = 0.5,
                                    do_plot = T) 
  
```


## 可视化遗传基因的映射
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
###可选择
gene<-"MAP2K2"
Pa<-"hsa04630"

data_mat<-Pagwas$data_mat[,rownames(Pagwas_fortify)]
pca_scCell_mat<-Pagwas$pca_scCell_mat[,rownames(Pagwas_fortify)]

Pagwas_fortify$gene<-as.vector(data_mat[gene,])
Pagwas_fortify$Pa<-as.vector(pca_scCell_mat[Pa,])

plot3 <-  ggplot() +
        geom_point(data = Pagwas_fortify,
                   aes(x = UMAP_1, y = UMAP_2,color =gene), size =0.6, alpha = 1) +
        umap_theme() +
        scale_colour_gradient2(low="#8479E1",mid="#F7F5F2",high="#FD5D5D")+
        theme(aspect.ratio=1) +
        guides(colour = guide_legend(override.aes = list(size=3)))+
        labs(title = gene)
#pdf("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_pruneGSE138852/umap_opc_scPagwas_p.pdf")
  print(plot3)
#  dev.off()
  
  plot4 <-  ggplot() +
        geom_point(data = Pagwas_fortify,
                   aes(x = UMAP_1, y = UMAP_2,color =Pa), size =0.6, alpha = 1) +
        umap_theme() +
        scale_colour_gradient2(low="#8479E1",mid="#F7F5F2",high="#FD5D5D")+
        theme(aspect.ratio=1) +
        guides(colour = guide_legend(override.aes = list(size=3)))+
        labs(title = Pa)
#pdf("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_pruneGSE138852/umap_opc_scPagwas_p.pdf")
  print(plot4)
```


## 可视化通路基因网络图
选择感兴趣的细胞类型和通路，进行网络可视化
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
suppressMessages(require("WGCNA"))
suppressMessages(require("patchwork"))
suppressMessages(require("tidygraph"))
suppressMessages(require("ggraph"))
suppressMessages(require("igraph"))
source("E:/RPakage/visulizeplotR/plot_pathway_contribution_network.R")
require("ggsci")
require("graphlayouts")

#check the objects
rownames(Pagwas$scPathways_rankPvalue)<-colnames(Pagwas$Pathway_sclm_results)
celltypes="OPC"
pathways<-rownames(Pagwas$scPathways_rankPvalue)[order(Pagwas$scPathways_rankPvalue[,"OPC"],decreasing = F)[1:5]]

pathways<-c("hsa00250","hsa04630","hsa04666","hsa00100","hsa00020")
#hsa00250 Pathway : Name: Alanine, aspartate and glutamate metabolism; Metabolism; Amino acid metabolism BRITE hierarchy.
#hsa04630:Name: JAK-STAT signaling pathway - Homo sapiens (human) Description: The Janus kinase/signal transducers and activators of transcription (JAK/STAT) pathway is one of a handful of pleiotropic cascades used to transduce a multitude of signals for development 
#hsa04666: hsa04666 Pathway : Name: Fc gamma R-mediated phagocytosis - Homo sapiens (human) Description: Phagocytosis plays an essential role in host-defense mechanisms through the uptake and destruction of infectious pathogens. Specialized cell types including macrophages, neutrophils, and monocytes take part in this process in higher organisms.
#hsa00100: Glycolysis, core module involving three-carbon compounds 
#hsa00020 Pathway : Name: Citrate cycle (TCA cycle) - Homo sapiens (human) Description: The citrate cycle (TCA cycle, Krebs cycle) is an important aerobic pathway for the final steps of the oxidation of carbohydrates and fatty acids. 
#extrafont::font_import()
library(extrafont)
loadfonts()
lapply(1:length(pathways),function(ss){
 
  pathway<-pathways[ss]
 # library(showtext)
 #fond_add('Arial Narrow','/Library/Fonts/ArialNarrow.tiff')
# pdf(paste0("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_pruneGSE138852/",pathway,"_pthway_network.pdf"))
plot_pathway_contribution_network(mat_datExpr=Single_data@assays$RNA@data,                                              gene_contribution=Pagwas$gene_heritability_correlation,
                                  Celltype_anno=Pagwas$Celltype_anno,
                                  Pathway_list=Pagwas$Pathway_list,
                                  celltypes="OPC",
                                  pathwat_select=pathway,
                                        pathways_thre=0.05,
                                        adjacency.power=4,
                                        n_max_genes=100,
                                        #edge_filter=0.00015,
                                        text_topn=10,
                                        igraph_algorithm = "drl",
                                        h_method = "complete",
                                        fontface_labels="bold.italic",
                                        color_edge = "#9D9D9D",
                                        color_group=pal_d3(alpha =0.5)(5)[ss],
                                        node_alpha=0.8,
                                        fontSize_label_lg=2,
                                        fontSize_legend_lg=3,
                                        fontSize_legend_xlg=3,
                                        limits=c(-0.1,0.1),
                                  range=c(1,7),
                                        edge_thickness = 2,
                                        do_plot=T , #figurenames=paste0("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_pruneGSE138852/",pathway,"_pthway_network.pdf"),
                                        width=10,
                                        height=10
  )
#dev.off()
})

```

## 可视化细胞类型遗传特异性通路-dotplot

```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
  library(tidyverse)
  library("rhdf5")
 library(ggplot2)
 library(grDevices)
 library(stats)
 library(FactoMineR)
 library(scales)
 library(reshape2)
 library(ggdendro)
 library(grImport2)
 library(gridExtra)
 library(grid)
 library(sisal)
source("E:/RPakage/visulizeplotR/plot_scpathway_contri_dot.R")

plot_scpathway_dot(Pagwas=Pagwas_seu,
                   celltypes=unique(Idents(Pagwas_seu))[1:5],
                             topn_path_celltype=5,
                             filter_p=0.05,
                             max_logp=15,
                             display_max_sizes=F,
                             size_var ="logrankPvalue" ,
                             col_var="proportion",
                             shape.scale = 8,
                             cols.use=c("lightgrey", "#E45826"),
                             dend_x_var = "logrankPvalue",
                             dist_method="euclidean",
                             hclust_method="ward.D",
                             do_plot = T,
                             figurenames = NULL,
                             width = 7,
                             height = 7)
                             
```

## 可视化基因遗传关联性排秩点图
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
source("E:/RPakage/visulizeplotR/heritability_cor_scatterplot.R")

heritability_cor_scatterplot(gene_heri_cor=Pagwas$gene_heritability_correlation,
                                       topn_genes_label=10,
                                       color_low="#035397",
                                       color_high ="#F32424",
                                       color_mid = "white",
                                       text_size=3,
                                       do_plot=T,
                                       #figurenames="E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_pruneGSE138852/test_scatterplot.pdf",
                                       width = 7,
                                       height = 7
)


```

### Plot the top5 heritability correlation genes in celltypes

```{r vln_Corgenes,message=FALSE,fig.height=6, fig.width=8,warning = FALSE, eval=FALSE}
top5genes<-rownames(Pagwas$gene_heritability_correlation)[order(Pagwas$gene_heritability_correlation,decreasing = T)[1:5]]
plot_vln_Corgenes(seurat_obj=Single_data,
             assay="RNA", slot="data",
             var_group="oupSample.cellType",
             vec_features=top5genes,
             vec_group_colors= pal_d3(alpha =0.5)(10),
             do_plot = T
             )

```

## 可视化细胞类型遗传关联显著性
```{r message=FALSE,results = "hide",warning = FALSE, eval=FALSE}
#pdf("E:/OneDrive/GWAS_Multiomics/ad_test/5.6test/ad_GSE138852/bar_celltypes.pdf")
Bootstrap_P_Barplot(p_results=Pagwas$bootstrap_results$bp_value[-1],
                                p_names=rownames(Pagwas$bootstrap_results)[-1],
                                title = "ad Test scPagwas",
                                figurenames = NULL,
                                width = 5,
                                height = 7,
                                do_plot=TRUE)
#dev.off()
Bootstrap_estimate_Plot(Pagwas=Pagwas,
                        figurenames = NULL,
                        width = 9,
                        height = 7,
                        do_plot=T)
```

## 拆分数据跑之后再合并
```{r, eval=FALSE}
 #1.start to run the wrapper functions for preprogress.
 Pagwas_data1<-scPagwas_main(Pagwas = NULL,
                     gwas_data =system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"),
                     Single_data ="E:/RPakage/scPagwas/inst/extdata/scRNAexample_split1.rds",
                     output.prefix="test1",
                     output.dirs="scPagwastest_output",
                     block_annotation = block_annotation,
                     assay="RNA",
                     Pathway_list=Genes_by_pathway_kegg,
                     chrom_ld = chrom_ld,
                     singlecell=T,
                     seruat_return=T,
                     celltype=T,
                     ncores = 1)
 
  Pagwas_data2<-scPagwas_main(Pagwas = NULL,
                     gwas_data =system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"),
                     Single_data ="E:/RPakage/scPagwas/inst/extdata/scRNAexample_split2.rds",
                     output.prefix="test2",
                     output.dirs="scPagwastest_output",
                     block_annotation = block_annotation,
                     assay="RNA",
                     Pathway_list=Genes_by_pathway_kegg,
                     chrom_ld = chrom_ld,
                     singlecell=T,
                     seruat_return=T,
                     celltype=T,
                     ncores = 1)
Pagwas_merge <- merge_pagwas(Pagwas_list = list(Pagwas_data1, Pagwas_data2))

```
