---
title: "my-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## 1.Adult-healthy-MDD
```{r setup, eval = FALSE}

#' scRNAexample
#'
#' @description
#' A single cell example data from \code{
#' liver_data<-readRDS("/share/pub/qiuf/COVID/liver/Rpoly.rds")
#' #set Idents
#' liver_data@meta.data$annotation <- liver_data@meta.data$anno
#' Idents(liver_data) <- liver_data@meta.data$anno
#' #normalization
#' liver_data <- NormalizeData(liver_data, normalization.method = "LogNormalize", scale.factor = 10000)
#' #correct the cell names, need no specific signal
#' anno <- liver_data@meta.data$annotation
#'  anno<-str_replace_all(anno,"-",".")
#'  anno<-str_replace_all(anno," ",".")
#' anno[which(anno=="Inflammatory.monocyte/macrophages")]<-"Inflammatory.monocyte.macrophages"
#' Celltype_anno$annotation<-str_replace_all(anno,"\\+",".")
#' Idents(liver_data) <- as.factor(anno)
#' liver_data<-subset(liver_data,idents = c("Stellate.cell","unknow","γδ.T.cell"), invert = TRUE)
#' scRNAexample<-liver_data[,sample(ncol(liver_data),1000)]
#'
#'  for (x in 1:length(unique(scRNAexample@meta.data$anno))){
#'  newnames <- paste0("Celltype",x)
#'  oldnames<-unique(scRNAexample@meta.data$anno)[x]
#'  scRNAexample@meta.data$anno[which(scRNAexample@meta.data$anno==oldnames)]<-newnames
#'  }
#'  Idents(scRNAexample)<-scRNAexample@meta.data$anno
#'  scRNAexample@meta.data<-scRNAexample@meta.data[,1:15]
#'
#'  save(scRNAexample,file = "E:/RPakage/scPagwas/data/scRNAexample.RData")
#'
#' }
#'
#' @docType data
#' @name scRNAexample
#' @usage data(scRNAexample)
#' @format a Seruat.
#' @source Generated from PBC
#' @keywords datasets
#' @examples
#' data(scRNAexample)
#' str(scRNAexample)
NULL




#' GWAS_summ_example
#'
#' @description
#' A GWAS summary data from \code{
#' library(readr)
#' library(dplyr)
#' GWAS_liver<-read_delim("/share/pub/mayl/Sherlock/XiangbingYu_Primary_biliary_cholangitis/PBC_GWAS_UKBiobank_summary_final",delim="\t")
#' #select the specific coloumn names: chrom，pos，rsid，beta，se，maf
#' GWAS_liver<-GWAS_liver[,c("CHR","POS","SNP","p","SE")]
#' colnames(GWAS_liver)<-c("chrom","pos","rsid","p","se")
#' COVOD19<-read_delim("/share/pub/mayl/01_COVID19_GWAS_round4/COVID19_HGI_B2_ALL_leave_23andme_20201020.b37.txt",delim="\t")
#' COVOD19<-COVOD19[,c("all_meta_AF","rsid")]
#' colnames(COVOD19)<-c("maf","rsid")
#' ##get beta
#' GWAS_liver$beta<-GWAS_liver$se*qchisq(GWAS_liver$p,1,lower.tail=F)
#' GWAS_liver_inner <- inner_join(GWAS_liver,COVOD19,by="rsid")
#' GWAS_liver_inner <- as.data.frame(GWAS_liver_inner)
#' GWAS_summ_example <- as.data.frame(GWAS_liver_inner[sample(nrow(GWAS_liver_inner),10000),])
#' save(GWAS_summ_example,file="/share/pub/dengcy/GWAS_Multiomics/pagwas/data/GWAS_summ_example.RData")
#' }
#'
#' @docType data
#' @name GWAS_summ_example
#' @usage data(GWAS_summ_example)
#' @format a data.frame.
#' @source Generated from PBC_GWAS_UKBiobank_summary_final
#' @keywords datasets
#' @examples
#' data(GWAS_summ_example)
#' str(GWAS_summ_example)
NULL




library("Seurat")
library("stringr") 
library(ggplot2)
library(ggthemes)
library(ggpubr)
library(scPagwas)
 library(SOAR)
options(future.globals.maxSize= 891289600)
data<-readRDS("/share/pub/qiuf/brain/01-data/healthy/adult/resolution0.2_final1.rds")
object.size(data)
#We advice you to set a R_LOCAL_CACHE for your data
Sys.setenv(R_LOCAL_CACHE="scPagwasSession")

object.size(data)
 data(Genes_by_pathway_kegg)

```

##R_LOCAL_CACHE
1.不同的函数如果同时运行要有不用的session命名，如果是连续运行则可以用同一个session文件
2.运行完代码之后如果需要删除当前环境中的变量，以免占用内存
查看当前session中的变量
Objects()
删除当前环境中的变量：
Remove(Objects())

```{r realexample, eval = FALSE}
 library(scPagwas)
 suppressMessages(library(Seurat))
 suppressMessages(library("dplyr"))
 data(Genes_by_pathway_kegg)
 #gene annotation files.
 data(block_annotation)
 #LD data
 data(chrom_ld)
 ##############真实例子：
  Pagwas<-scPagwas_main(Pagwas = NULL,
                     gwas_data ="E:/OneDrive/GWAS_Multiomics/ad_test/AD_prune_gwas_data.txt",
                     add_eqtls="OnlyTSS",
                     block_annotation = block_annotation,
                     assay="RNA",
                     Single_data = "E:/OneDrive/GWAS_Multiomics/ad_test/AD_final.rds",
                     nfeatures =2000,
                     Pathway_list=Genes_by_pathway_kegg,
                     chrom_ld = chrom_ld)
  
  #######################分步骤查看结果
  Pagwas<-Celltype_heritability_contributions(Pagwas,iters = 200)

  
 Bootstrap_P_Barplot(Pagwas=Pagwas,
                    figurenames = NULL,
                    width = 5,
                    height = 7,
                    do_plot=T,
                    title = "AD scPagwas")
Pagwas<-Singlecell_heritability_contributions(Pagwas,
                                              n.cores=1,
                                              part = 0.5,
                                              SimpleResult=F)
################################
library(ggplot2)
library(grDevices)
library(stats)
library(FactoMineR)
library(scales)
library(reshape2)
library(ggdendro)
library(grImport2)
library(gridExtra)
library(grid)
library(sisal)

#load("E:/OneDrive/GWAS_Multiomics/dox/FlexDotPlot-master/FlexDotPlot-master/data/PBMC3K_example_data.rda")
setwd("E:/OneDrive/GWAS_Multiomics/ad_test")
load("pagwas_ad_test.RData")
Pagwas<-pagwas
rm(pagwas)
Pagwas$pca_scCell_mat<- apply(Pagwas$pca_scCell_mat,1, function(x) (x - min(x)) / (max(x) - min(x)))

pca_scCell_mat<-Pagwas$pca_scCell_mat[,colnames(Pagwas$Pathway_sclm_results)]

Pagwas$Pathway_sclm_mat<-Pagwas$Pathway_sclm_results * pca_scCell_mat
rownames(Pagwas$Pathway_sclm_mat)<-rownames(pca_scCell_mat)
colnames(Pagwas$Pathway_sclm_mat)<-colnames(Pagwas$pca_scCell_mat)

Pagwas$Celltype_anno$annotation<-as.factor(Pagwas$Celltype_anno$annotation)

celltype_pathway_scproportion<-tapply(Pagwas$Celltype_anno$cellnames,Pagwas$Celltype_anno$annotation,function(x){
  #x[1:3]
  return(colMeans(Pagwas$Pathway_sclm_mat[x,]))
})

celltype_pathway_scproportion<-Reduce(function(dtf1, dtf2) rbind(dtf1, dtf2),celltype_pathway_scproportion)

#celltype_pathway_scproportion<-as.data.frame(apply(celltype_pathway_scproportion,2, function(x) (x - min(x)) / (max(x) - min(x))))

rownames(celltype_pathway_scproportion)<-levels(Pagwas$Celltype_anno$annotation)

Pathway_sclm_mat<-as(Pagwas$Pathway_sclm_mat,"matrix")


ls_scCell<-CreateSeuratObject(
  t(Pathway_sclm_mat),
  project = "CreateSeuratObject",
  assay = "RNA",
  names.field = 1,
  names.delim = "_",
  meta.data = Pagwas$Celltype_anno
)

Idents(ls_scCell)<-Pagwas$Celltype_anno$annotation
#ls_scCell_markerpa <- FindAllMarkers(object = ls_scCell)
ls_scCell <- FindVariableFeatures(object = ls_scCell,nfeatures = dim(ls_scCell)[1]*0.1)

#pa_cellytpes<-lapply(unique(ls_scCell_markerpa$cluster),function(x){
#  ls_scCell_markerpa[ls_scCell_markerpa$cluster==x,"gene"][1:3]
#})

pca_scCell<-CreateSeuratObject(
  t(pca_scCell_mat),
  project = "CreateSeuratObject",
  assay = "RNA",
  names.field = 1,
  names.delim = "_",
  meta.data = Pagwas$Celltype_anno
)

Idents(pca_scCell)<-Pagwas$Celltype_anno$annotation
pca_scCell <- FindVariableFeatures(object = pca_scCell,nfeatures = dim(pca_scCell)[1]*0.1)
#pca_scCell_markerpa <- FindAllMarkers(object = pca_scCell,min.pct = 0,logfc.threshold = 0,return.thresh = 0.1)

#pas<-VariableFeatures(object = pca_scCell)
pas1<-VariableFeatures(object = pca_scCell)
pas2<-VariableFeatures(object = ls_scCell)

#pas<-VariableFeatures(object = pca_scCell)

pas<-names(sort(Pagwas$Pathway_block_heritability,decreasing = T)[1:35])
a1<-as.data.frame(celltype_pathway_scproportion[,pas])
a2<-as.data.frame(t(Pagwas$pca_cell_df))[,pas]

a1$celltype<-rownames(a1)
a2$celltype<-rownames(a2)

gg_pa<-reshape2::melt(a1,id.vars="celltype",variable.name = "pathway",value.name = "heritability_proporion")

gg_pa2<-reshape2::melt(a2,id.vars="celltype",variable.name = "pathway",value.name = "pca")
gg_pa<-merge(gg_pa,gg_pa2)
gg_pa<-gg_pa[,c("pathway","celltype","heritability_proporion","pca")]

pdf("E:/OneDrive/GWAS_Multiomics/ad_test/AD_plot_sc_pathway.pdf",width = 13)
plot_sc_pa_contri_dot(data.to.plot = gg_pa, 
                      size_var = "pca", 
                      col_var="heritability_proporion",
                      shape.scale = 8,
                      #cols.use="#E45826",
                      dend_x_var = c("heritability_proporion"),
                      dist_method="euclidean", 
                      hclust_method="ward.D"
)
dev.off()
lapply(names(sort(Pagwas$Pathway_block_heritability,decreasing = T))[c(32,33,35)],function(pa){
  print(pa)
  intersect(Genes_by_pathway_kegg[[pa]],c("CREB1","GSK3B"))

})
#intersect(Genes_by_pathway_kegg$hsa05322,c("CREB1","GSK3B"))

```
