---
title: "Split_big_scRNAseqData_and_integrate_result"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Split_big_scRNAseqData_and_integrate_result}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
 collapse = TRUE,
 comment = "#>",
  fig.path = "figures/vignette-",
 out.width = "60%"
)
```

```{r setup}
library(scPagwas)
```

## Run the scPagwas for big scRNA-seq data 
`If your computer storage is enough, ignore this item`.
Sometimes, the storage is limit for your analysis, you can splice the scRNA-seq data set by celltypes or in order. 

scPagwas don't limit the way to cut the data set, but we advice to do it by cell types, because some clusters may be filtered for the minimum cell number. 


### 1. Splice scRNA-seq data 
First, we splice the big scRNA-seq data for several sub-datas by cell types.

```{r}
library(Seurat)
scRNAexample <-readRDS(system.file("extdata", "scRNAexample.rds", package = "scPagwas"))
Example_splice1<-subset(scRNAexample,idents=c("Celltype1","Celltype2","Celltype3","Celltype4", "Celltype5"))
Example_splice2<-subset(scRNAexample,idents=c( "Celltype6","Celltype7","Celltype8","Celltype9","Celltype10" ))
```

### 2. Run gwas data input 

Second, we run the `GWAS_summary_input` and `Snp2Gene` firstly, because the following function share the same gwas data.
Sometimes, we need not run these functions for small gwas data, because a little time for these functions.

```{r}
Pagwas<-list()
gwas_data <- bigreadr::fread2(system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"))
Pagwas <- GWAS_summary_input(
    Pagwas = Pagwas,
    gwas_data = gwas_data,
    maf_filter = 0.1
  )
Pagwas$snp_gene_df <- Snp2Gene(snp = Pagwas$gwas_data, refGene = block_annotation, marg = 10000)
names(Pagwas)
```

### 2. Run scPagwas for sub-data

```{r echo=T,results='hide',message=FALSE, warning=FALSE}
Pagwas1<-scPagwas_main(Pagwas =Pagwas,
                     gwas_data =NULL,
                     Single_data =Example_splice1,
                     output.prefix="splice1",
                     output.dirs="splice_scPagwas",
                     Pathway_list=Genes_by_pathway_kegg,
                     ncores=5,
                     assay="RNA",
                     block_annotation = block_annotation,
                     chrom_ld = chrom_ld)
if(length(SOAR::Objects())>0){
 SOAR::Remove(SOAR::Objects()) 
}
# ##------ Thu Jul 14 12:48:37 2022 ------## ******* 1st: Single_data_input function start! ********
# Input single cell data!
# 5 cell types are remain, after filter!
# done!
# ##------ Thu Jul 14 12:48:40 2022 ------## ******* 2nd: Pathway_pcascore_run function start!! ********
# * Start to get Pathway SVD socre!
#   |=================                                                                    |  20%[1] "Celltype1"
#   |==================================                                                   |  40%[1] "Celltype2"
#   |===================================================                                  |  60%[1] "Celltype3"
#   |====================================================================                 |  80%[1] "Celltype4"
#   |=====================================================================================| 100%[1] "Celltype5"
# 
# Using score as value column: use value.var to override.
# done!
# ##------ Thu Jul 14 12:48:45 2022 ------## ******* 3rd: GWAS_summary_input function start! ********
# ##------ Thu Jul 14 12:48:45 2022 ------## ******* 5th: Pathway_annotation_input function start! ********
# obtain the pathway block information
# done!
# ##------ Thu Jul 14 12:48:47 2022 ------## ******* 6th: Link_pathway_blocks_gwas function start! ********
# * Start to link gwas and pathway block annotations for 278 pathways!
#   |=====================================================================================| 100%
# done!
# ##------ Thu Jul 14 12:49:12 2022 ------## ******* 7th: Celltype_heritability_contributions function start! ********
# ** Start inference
# * starting bootstrap iteration for 200 times
#   |=====================================================================================| 100%
# done!
# ##------ Thu Jul 14 12:49:14 2022 ------## ******* 8th: scPagwas_perform_score function start! ********
# * Get Pathways'rankPvalue for each celltypes!
# * Get scPgwas score for each single cell
# done!
# ##------ Thu Jul 14 12:49:19 2022 ------## ******* 9th: scGet_gene_heritability_correlation function start! ********
# done
# * Get rankPvalue for each single cell
```

```{r echo=T,results='hide',message=FALSE, warning=FALSE}
Pagwas2<-scPagwas_main(Pagwas =Pagwas,
                     gwas_data =NULL,
                     Single_data =Example_splice2,
                     output.prefix="splice2 ",
                     output.dirs="splice_scPagwas",
                     Pathway_list=Genes_by_pathway_kegg,
                     ncores=5,
                     assay="RNA",
                     block_annotation = block_annotation,
                     chrom_ld = chrom_ld)
if(length(SOAR::Objects())>0){
 SOAR::Remove(SOAR::Objects()) 
}

# ##------ Thu Jul 14 11:19:59 2022 ------## ******* 1st: Single_data_input function start! ********
# Input single cell data!
# 5 cell types are remain, after filter!
# done!
# ##------ Thu Jul 14 11:20:00 2022 ------## ******* 2nd: Pathway_pcascore_run function start!! ********
# * Start to get Pathway SVD socre!
#   |=================                                                                     |  20%[1] "Celltype6"
#   |==================================                                                    |  40%[1] "Celltype7"
#   |====================================================                                  |  60%[1] "Celltype8"
#   |=====================================================================                 |  80%[1] "Celltype9"
#   |======================================================================================| 100%[1] "Celltype10"
# 
# Using score as value column: use value.var to override.
# done!
# ##------ Thu Jul 14 11:20:06 2022 ------## ******* 3rd: GWAS_summary_input function start! ********
# ##------ Thu Jul 14 11:20:06 2022 ------## ******* 5th: Pathway_annotation_input function start! ********
# obtain the pathway block information
# done!
# ##------ Thu Jul 14 11:20:07 2022 ------## ******* 6th: Link_pathway_blocks_gwas function start! ********
# * Start to link gwas and pathway block annotations for 273 pathways!
#   |======================================================================================| 100%
# done!
# ##------ Thu Jul 14 11:20:22 2022 ------## ******* 7th: Celltype_heritability_contributions function start! ********
# ** Start inference
# * starting bootstrap iteration for 200 times
#   |======================================================================================| 100%
# done!
# ##------ Thu Jul 14 11:20:24 2022 ------## ******* 8th: scPagwas_perform_score function start! ********
# * Get Pathways'rankPvalue for each celltypes!
# * Get scPgwas score for each single cell
# done!
# ##------ Thu Jul 14 11:20:28 2022 ------## ******* 9th: scGet_gene_heritability_correlation function start! ********
# done
# * Get rankPvalue for each single cell
```

### 3. Integrate the result for splice data. 

We integrate the splice result by `merge_pagwas` function. 

```{r}
Pagwas_integrate <- merge_pagwas(Pagwas_list = list(Pagwas1,Pagwas2),n_topgenes = 1000)
print(Pagwas_integrate)
```
