<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="scPagwas"><title>Data_input_preproccess_for_scPagwas • scPagwas</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data_input_preproccess_for_scPagwas"><meta property="og:description" content="scPagwas"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scPagwas</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.10.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Bmmc_monocytecount_singlecell_celltype_vignette.html">Bmmc_monocytecount_singlecell_celltype_vignette</a>
    <a class="dropdown-item" href="../articles/Data_input_preproccess_for_scPagwas.html">Data_input_preproccess_for_scPagwas</a>
    <a class="dropdown-item" href="../articles/Multi_TraitFiles_for_one_SingleData_running.html">Multi_TraitFiles_for_one_SingleData_running</a>
    <a class="dropdown-item" href="../articles/Prune_GwasSummaryFIles_plink.html">Prune_GwasSummaryFIles_plink</a>
    <a class="dropdown-item" href="../articles/Running_scPagwas_steps_by_SubFunctions.html">Running_scPagwas_steps_by_SubFunctions</a>
    <a class="dropdown-item" href="../articles/Split_big_scRNAseqData_and_integrate_result.html">Split_big_scRNAseqData_and_integrate_result</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Data_input_preproccess_for_scPagwas</h1>
            
            <h4 data-toc-skip class="date">Last Updated: 29, 九月, 2022
at 14:52</h4>
      
      
      <div class="d-none name"><code>Data_input_preproccess_for_scPagwas.Rmd</code></div>
    </div>

    
    
<div id="data-input" class="section level1">
<h1>Data input</h1>
<p>There were three input resources for scPagwas: scRNA-seq
dataset(seruat format), a GWAS summary dataset(txt file) on a given
phenotype and an extensive panel of pathways or functional gene
sets(gene symbol list).</p>
<div id="single-cell-data-input" class="section level2">
<h2>1.Single cell data Input</h2>
<p>The healthy BMMC scRNA-seq dataset was downloaded from the website:
<a
href="https://jeffgranja.s3.amazonaws.com/MPAL-10x/Supplementary_Data/Healthy-Data/scRNA-Healthy-Hematopoiesis-191120.rds"
class="uri">https://jeffgranja.s3.amazonaws.com/MPAL-10x/Supplementary_Data/Healthy-Data/scRNA-Healthy-Hematopoiesis-191120.rds</a></p>
<ul>
<li>1). The scRNA-seq dataset should have Idents for cell types or
clusters.</li>
<li>2). The scRNA-seq dataset should be normalized.</li>
</ul>
<div id="read-the-scrna-seq-datasetseruat-format"
class="section level3">
<h3>1.1.read the scRNA-seq dataset(seruat format)</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;scPagwas&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;scRNAseq&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;SingleCellExperiment&quot;</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;stringr&quot;</span>) </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>scRNA_Healthy_Hema<span class="ot">&lt;-</span><span class="fu">readRDS</span>(<span class="st">&quot;E:/OneDrive/SingleCell/data/PBMCscATAC-seq/scRNA-Healthy-Hematopoiesis-191120.rds&quot;</span>)</span></code></pre></div>
<p>Because the Example data <code>scRNA_Healthy_Hema</code> is
SingleCellExperiment format, here creating a seruat format data:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(scRNA_Healthy_Hema, <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>Seu_Healthy_Hema <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">counts =</span> counts, </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">meta.data=</span><span class="fu">as.data.frame</span>(<span class="fu">colData</span>(scRNA_Healthy_Hema)),</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">min.cells =</span> <span class="dv">3</span>, </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">min.features =</span> <span class="dv">200</span>)</span></code></pre></div>
</div>
<div id="set-idents" class="section level3">
<h3>1.2.Set Idents</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(Seu_Healthy_Hema)<span class="ot">&lt;-</span>scRNA_Healthy_Hema<span class="sc">@</span>colData<span class="sc">$</span>BioClassification</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">Idents</span>(Seu_Healthy_Hema))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>        01_HSC 02_Early.Eryth  03_Late.Eryth  04_Early.Baso    05_CMP.LMPP       06_CLP<span class="fl">.1</span> </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>          <span class="dv">1425</span>           <span class="dv">1653</span>            <span class="dv">446</span>            <span class="dv">111</span>           <span class="dv">2260</span>            <span class="dv">903</span> </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>        07_GMP    08_GMP.Neut         09_pDC         10_cDC 11_CD14.Mono<span class="fl">.1</span> 12_CD14.Mono<span class="fl">.2</span> </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>          <span class="dv">2097</span>           <span class="dv">1050</span>            <span class="dv">544</span>            <span class="dv">325</span>           <span class="dv">1800</span>           <span class="dv">4222</span> </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  13_CD16.Mono         14_Unk       15_CLP<span class="fl">.2</span>       16_Pre.B           17_B      18_Plasma </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>           <span class="dv">292</span>            <span class="dv">520</span>            <span class="dv">377</span>            <span class="dv">710</span>           <span class="dv">1711</span>             <span class="dv">62</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>      19_CD8.N      20_CD4.N1      21_CD4.N2       22_CD4.M      23_CD8.EM      24_CD8.CM </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>          <span class="dv">1521</span>           <span class="dv">2470</span>           <span class="dv">2364</span>           <span class="dv">3539</span>            <span class="dv">796</span>           <span class="dv">2080</span> </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>         25_NK         26_Unk </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>          <span class="dv">2143</span>            <span class="dv">161</span> </span></code></pre></div>
</div>
<div id="normalized" class="section level3">
<h3>1.3.Normalized</h3>
<p>Scale and Normalize the dataset.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>Seu_Healthy_Hema <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(Seu_Healthy_Hema)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>Seu_Healthy_Hema <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(Seu_Healthy_Hema, <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>)</span></code></pre></div>
<p>The BMMC processed scRNA-seq data can download from here: <a
href="https://1drv.ms/u/s!As-aKqXDnDUHi9pNciEgQFbe-AHgLw?e=3XsNnQ"
class="uri">https://1drv.ms/u/s!As-aKqXDnDUHi9pNciEgQFbe-AHgLw?e=3XsNnQ</a></p>
</div>
</div>
<div id="gwas-summary-data-input" class="section level2">
<h2>2.GWAS summary data Input</h2>
<p>GWAS Summary statistics are download from: <a
href="ftp://ftp.sanger.ac.uk/pub/project/humgen/summary_statistics/UKBB_blood_cell_traits/"
class="uri">ftp://ftp.sanger.ac.uk/pub/project/humgen/summary_statistics/UKBB_blood_cell_traits/</a>
for UK Biobank.</p>
<ul>
<li>basophil cell count: ieu-b-29</li>
<li>eosinophil cell count : ieu-b-33</li>
<li>lymphocyte cell count: ebi-a-GCST004627</li>
<li>monocyte cell count: ieu-b-31</li>
<li>neutrophil cell count: ieu-b-34</li>
<li>white blood cell count: ieu-b-30</li>
<li>Lymphocyte percentage of white cells: ebi-a-GCST004632</li>
<li>Hemoglobin concentration: ebi-a-GCST90002311</li>
<li>Mean corpuscular hemoglobin concentration: ebi-a-GCST90002329</li>
<li>Mean corpuscular volume: ebi-a-GCST90002335</li>
</ul>
<p>The GWAS Summary statistics file need to be processed into a “txt”
file including six coloumn and tab-delimited.</p>
<div id="check-the-example-gwas-summary-statistics-file"
class="section level3">
<h3>2.1.Check the example GWAS Summary statistics file</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scPagwas)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>gwas_data <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;GWAS_summ_example.txt&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scPagwas&quot;</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(gwas_data))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">chrom</th>
<th align="right">pos</th>
<th align="left">rsid</th>
<th align="right">se</th>
<th align="right">beta</th>
<th align="right">maf</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">6</td>
<td align="right">119968580</td>
<td align="left">rs1159767</td>
<td align="right">0.0329339</td>
<td align="right">0.0193056</td>
<td align="right">0.5275</td>
</tr>
<tr class="even">
<td align="right">10</td>
<td align="right">130566523</td>
<td align="left">rs559109</td>
<td align="right">0.0339840</td>
<td align="right">0.0454866</td>
<td align="right">0.4047</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">133328825</td>
<td align="left">rs6893145</td>
<td align="right">0.0486105</td>
<td align="right">0.1447440</td>
<td align="right">0.1222</td>
</tr>
<tr class="even">
<td align="right">7</td>
<td align="right">146652932</td>
<td align="left">rs13228798</td>
<td align="right">0.0350390</td>
<td align="right">0.0039275</td>
<td align="right">0.3211</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="right">23373027</td>
<td align="left">rs4400806</td>
<td align="right">0.0323868</td>
<td align="right">0.0108843</td>
<td align="right">0.4604</td>
</tr>
<tr class="even">
<td align="right">5</td>
<td align="right">168561511</td>
<td align="left">rs1421766</td>
<td align="right">0.0324148</td>
<td align="right">0.0260761</td>
<td align="right">0.5182</td>
</tr>
</tbody>
</table>
</div>
<div
id="take-the-monocyte-count-as-real-example-and-deal-with-it-step-by-step"
class="section level3">
<h3>2.2.Take the monocyte count as real example and deal with it step by
step:</h3>
<p>Preprogress the raw monocytecount GWAS Summary statistics file.
Extract the useful information from raw files. Download the Raw gwas
GWAS Summary statistics file: <a
href="https://gwas.mrcieu.ac.uk/files/ieu-b-31/ieu-b-31.vcf.gz"
class="uri">https://gwas.mrcieu.ac.uk/files/ieu-b-31/ieu-b-31.vcf.gz</a></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="do">############read the raw vcf file</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>monocytecount<span class="ot">&lt;-</span><span class="st">&quot;ieu-b-31.vcf.gz&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>GWAS_raw <span class="ot">&lt;-</span><span class="fu">read_table2</span>(monocytecount,<span class="at">comment =</span> <span class="st">&quot;#&quot;</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GWAS_raw )<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;CHROM&quot;</span>,<span class="st">&quot;POS&quot;</span>,<span class="st">&quot;ID&quot;</span>,<span class="st">&quot;REF&quot;</span>,<span class="st">&quot;ALT&quot;</span>,<span class="st">&quot;QUAL&quot;</span>,<span class="st">&quot;FILTER&quot;</span>,<span class="st">&quot;INFO&quot;</span>,<span class="st">&quot;FORMAT&quot;</span>,<span class="st">&quot;IEU&quot;</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#ES:SE:LP:AF:SS:ID</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">#ieu-a type of file:split the FORMAT and IEU coloumn, which including different information.</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>value_list<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(GWAS_raw),<span class="cf">function</span>(i){</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  index <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(GWAS_raw<span class="sc">$</span>FORMAT[i], <span class="at">split =</span> <span class="st">&quot;:&quot;</span>,<span class="at">fixed=</span>T))</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  value <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(GWAS_raw<span class="sc">$</span>IEU[i], <span class="at">split =</span> <span class="st">&quot;:&quot;</span>,<span class="at">fixed=</span>T))</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(value)<span class="ot">&lt;-</span>index</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  value<span class="ot">&lt;-</span>value[<span class="fu">c</span>(<span class="st">&quot;ES&quot;</span>,<span class="st">&quot;SE&quot;</span>,<span class="st">&quot;LP&quot;</span>,<span class="st">&quot;AF&quot;</span>,<span class="st">&quot;SS&quot;</span>)]</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(value)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a> value_df<span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">as.numeric</span>(<span class="fu">unlist</span>(value_list)),<span class="at">byrow=</span>T,<span class="at">ncol=</span><span class="dv">5</span>)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a> value_df<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(value_df)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a> <span class="fu">colnames</span>(value_df)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;ES&quot;</span>,<span class="st">&quot;SE&quot;</span>,<span class="st">&quot;LP&quot;</span>,<span class="st">&quot;AF&quot;</span>,<span class="st">&quot;SS&quot;</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a> gwas_data<span class="ot">&lt;-</span>GWAS_raw[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a> gwas_data<span class="ot">&lt;-</span> <span class="fu">cbind</span>(gwas_data,value_df)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a> <span class="fu">colnames</span>(gwas_data)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="co">#[1] &quot;CHROM&quot; &quot;POS&quot;   &quot;ID&quot;    &quot;REF&quot;   &quot;ALT&quot;   &quot;ES&quot;    &quot;SE&quot;    &quot;LP&quot;    &quot;AF&quot;   </span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="co">#[10] &quot;SS&quot;</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="co">#change the colnames:</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(gwas_data)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>,<span class="st">&quot;pos&quot;</span>,<span class="st">&quot;rsid&quot;</span>,<span class="st">&quot;REF&quot;</span>,<span class="st">&quot;ALT&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;se&quot;</span>,<span class="st">&quot;maf&quot;</span>,<span class="st">&quot;N&quot;</span>)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="co">#select the useful information for scPagwas.</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>gwas_data<span class="ot">&lt;-</span>gwas_data[,<span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>,<span class="st">&quot;pos&quot;</span>,<span class="st">&quot;rsid&quot;</span>,<span class="st">&quot;beta&quot;</span>,<span class="st">&quot;se&quot;</span>,<span class="st">&quot;maf&quot;</span>)]</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(gwas_data,<span class="at">file=</span><span class="st">&quot;monocytecount_gwas_data.txt&quot;</span>,<span class="at">row.names=</span>F,<span class="at">quote=</span>F)</span></code></pre></div>
</div>
</div>
<div id="pathway-gene-list" class="section level2">
<h2>3.Pathway gene list</h2>
<p>There are some processed pathway gene list provided by scPagwas
package: all these pathway list are including gene symbols.</p>
<ul>
<li>Genes_by_pathway_kegg</li>
<li>genes.by.celltype.pathway</li>
<li>genes.by.gpbp.pathway</li>
<li>genes.by.reactome.pathway</li>
<li>genes.by.regulatory.pathway</li>
<li>genes.by.tft.pathway</li>
<li>genes.by.hallmark.pathway</li>
<li>genes.by.immunologic.pathway</li>
<li>genes.by.immunesigdb.pathway</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>Genes_by_pathway_kegg[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $hsa00010</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;HK3&quot;     &quot;HK1&quot;     &quot;HK2&quot;     &quot;HKDC1&quot;   &quot;GCK&quot;     &quot;GPI&quot;     &quot;PFKM&quot;   </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [8] &quot;PFKP&quot;    &quot;PFKL&quot;    &quot;FBP1&quot;    &quot;FBP2&quot;    &quot;ALDOC&quot;   &quot;ALDOA&quot;   &quot;ALDOB&quot;  </span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [15] &quot;TPI1&quot;    &quot;GAPDH&quot;   &quot;GAPDHS&quot;  &quot;PGK2&quot;    &quot;PGK1&quot;    &quot;PGAM1&quot;   &quot;PGAM2&quot;  </span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [22] &quot;PGAM4&quot;   &quot;ENO3&quot;    &quot;ENO2&quot;    &quot;ENO1&quot;    &quot;ENO4&quot;    &quot;PKM&quot;     &quot;PKLR&quot;   </span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [29] &quot;PDHA2&quot;   &quot;PDHA1&quot;   &quot;PDHB&quot;    &quot;DLAT&quot;    &quot;DLD&quot;     &quot;LDHAL6A&quot; &quot;LDHAL6B&quot;</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [36] &quot;LDHA&quot;    &quot;LDHB&quot;    &quot;LDHC&quot;    &quot;ADH1A&quot;   &quot;ADH1B&quot;   &quot;ADH1C&quot;   &quot;ADH7&quot;   </span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [43] &quot;ADH4&quot;    &quot;ADH5&quot;    &quot;ADH6&quot;    &quot;AKR1A1&quot;  &quot;ALDH2&quot;   &quot;ALDH3A2&quot; &quot;ALDH1B1&quot;</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [50] &quot;ALDH7A1&quot; &quot;ALDH9A1&quot; &quot;ALDH3B1&quot; &quot;ALDH3B2&quot; &quot;ALDH3A1&quot; &quot;ACSS1&quot;   &quot;ACSS2&quot;  </span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [57] &quot;GALM&quot;    &quot;PGM1&quot;    &quot;PGM2&quot;    &quot;G6PC1&quot;   &quot;G6PC2&quot;   &quot;G6PC3&quot;   &quot;ADPGK&quot;  </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [64] &quot;BPGM&quot;    &quot;MINPP1&quot;  &quot;PCK1&quot;    &quot;PCK2&quot;   </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $hsa00020</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;CS&quot;     &quot;ACLY&quot;   &quot;ACO2&quot;   &quot;ACO1&quot;   &quot;IDH1&quot;   &quot;IDH2&quot;   &quot;IDH3B&quot;  &quot;IDH3G&quot; </span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [9] &quot;IDH3A&quot;  &quot;OGDHL&quot;  &quot;OGDH&quot;   &quot;DLST&quot;   &quot;DLD&quot;    &quot;SUCLG1&quot; &quot;SUCLG2&quot; &quot;SUCLA2&quot;</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [17] &quot;SDHA&quot;   &quot;SDHB&quot;   &quot;SDHC&quot;   &quot;SDHD&quot;   &quot;FH&quot;     &quot;MDH1&quot;   &quot;MDH2&quot;   &quot;PC&quot;    </span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [25] &quot;PCK1&quot;   &quot;PCK2&quot;   &quot;PDHA2&quot;  &quot;PDHA1&quot;  &quot;PDHB&quot;   &quot;DLAT&quot;  </span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $hsa00030</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;GPI&quot;     &quot;G6PD&quot;    &quot;PGLS&quot;    &quot;H6PD&quot;    &quot;PGD&quot;     &quot;RPE&quot;     &quot;RPEL1&quot;  </span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [8] &quot;TKT&quot;     &quot;TKTL2&quot;   &quot;TKTL1&quot;   &quot;TALDO1&quot;  &quot;RPIA&quot;    &quot;DERA&quot;    &quot;RBKS&quot;   </span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [15] &quot;PGM1&quot;    &quot;PGM2&quot;    &quot;PRPS1L1&quot; &quot;PRPS2&quot;   &quot;PRPS1&quot;   &quot;RGN&quot;     &quot;IDNK&quot;   </span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [22] &quot;GLYCTK&quot;  &quot;ALDOC&quot;   &quot;ALDOA&quot;   &quot;ALDOB&quot;   &quot;FBP1&quot;    &quot;FBP2&quot;    &quot;PFKM&quot;   </span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [29] &quot;PFKP&quot;    &quot;PFKL&quot;</span></span></code></pre></div>
<p>There are some steps for get these pathway gene list: There is no
need to repeat running these code.</p>
<div id="kegg-pathway-list" class="section level3">
<h3>KEGG pathway list</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(KEGGREST)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>pathways.list <span class="ot">&lt;-</span> <span class="fu">keggList</span>(<span class="st">&quot;pathway&quot;</span>, <span class="st">&quot;hsa&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Pull all genes for each pathway</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>pathway.codes <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;path:&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">names</span>(pathways.list)) </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>genes.by.pathway_kegg <span class="ot">&lt;-</span> <span class="fu">sapply</span>(pathways.list,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                           <span class="cf">function</span>(pwid){</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                             pw <span class="ot">&lt;-</span> <span class="fu">keggGet</span>(pwid)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                             <span class="cf">if</span> (<span class="fu">is.null</span>(pw[[<span class="dv">1</span>]]<span class="sc">$</span>GENE)) <span class="fu">return</span>(<span class="cn">NA</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                             pw2 <span class="ot">&lt;-</span> pw[[<span class="dv">1</span>]]<span class="sc">$</span>GENE[<span class="fu">c</span>(<span class="cn">FALSE</span>,<span class="cn">TRUE</span>)] <span class="co"># may need to modify this to c(FALSE, TRUE) for other organisms</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                             pw2 <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(pw2, <span class="at">split =</span> <span class="st">&quot;;&quot;</span>, <span class="at">fixed =</span> T), <span class="cf">function</span>(x)x[<span class="dv">1</span>]))</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">return</span>(pw2)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                           })</span></code></pre></div>
</div>
<div id="other-pathway-list" class="section level3">
<h3>Other pathway list</h3>
<p>These pathways data are download from <a
href="https://www.gsea-msigdb.org/gsea/index.jsp">GSEA</a> website.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c8.all.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>genes.by.celltype.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c3.tft.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>genes.by.tft.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c5.go.bp.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>genes.by.gpbp.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c3.all.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>genes.by.regulatory.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;h.all.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>genes.by.hallmark.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c2.cp.reactome.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>genes.by.reactome.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(genes.by.reactome.pathway,<span class="at">file =</span> <span class="st">&quot;genes.by.reactome.pathway.RData&quot;</span>)</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a><span class="co">#all immunologic signature gene sets as Gene Symbols</span></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c7.all.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>genes.by.immunologic.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a><span class="co">#ImmuneSigDB gene sets as Gene Symbols</span></span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c7.immunesigdb.v7.5.1.symbols.gmt&quot;</span>)</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>genes.by.immunesigdb.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span></code></pre></div>
<p>Note:</p>
<ol style="list-style-type: decimal">
<li>Sometimes, there is no need to change the pathway list frequently
once you choose a fitable pathway list.</li>
<li>The summed number of genes in pathway list should not too much to
cost a long time or memory.Once your pathway list is too big, you can
choose some sub-pathway suitable for you analysisi or remove some
redundance pathways.</li>
</ol>
</div>
<div id="reduce-pathway-gene-list" class="section level3">
<h3>Reduce pathway gene list</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">length</span>(genes.by.reactome.pathway)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1615</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">length</span>(<span class="fu">unlist</span>(genes.by.reactome.pathway))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 89476</span></span></code></pre></div>
<p>We set the <code>genes.by.reactome.pathway</code> for example, there
are 1615 gene list and 89476 genes, a mount of resouces and time will be
cost when running scPagwas, we provid <code>reduce_pathway</code> to
reduce the pathway list:</p>
<ul>
<li><code>pathway_seed</code> choose a list of pathway names as seed,
which are some pahtways cann’t be remove.</li>
<li><code>remove_proporion</code> The propotion of duplicated between
seed pathway and the others</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>reduce_genes.by.reactome.pathway<span class="ot">&lt;-</span>scPagwas<span class="sc">::</span><span class="fu">reduce_pathway</span>(<span class="at">pathway_seed=</span><span class="fu">names</span>(genes.by.reactome.pathway)[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes.by.reactome.pathway),<span class="dv">500</span>)],</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">pathway_list=</span>genes.by.reactome.pathway,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">remove_proporion=</span><span class="fl">0.6</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(reduce_genes.by.reactome.pathway)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#[1] 1045</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unlist</span>(reduce_genes.by.reactome.pathway))</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#[1] 34980</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="do">##C7: immunologic signature gene sets</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">&quot;c7.all.v2022.1.Hs.symbols.gmt&quot;</span>)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">vapply</span>(res, <span class="cf">function</span>(y) y[<span class="dv">1</span>], <span class="fu">character</span>(<span class="dv">1</span>))</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>genes.by.immunologic.pathway <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="st">&quot;[&quot;</span>, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="do">###The length of Pathway will not lager than 1000 and not smaller than 100.</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>reduce_genes.by.immunologic.pathway<span class="ot">&lt;-</span>genes.by.immunologic.pathway[<span class="fu">unlist</span>(<span class="fu">lapply</span>(genes.by.immunologic.pathway,<span class="cf">function</span>(x) <span class="fu">length</span>(x)<span class="sc">&gt;</span><span class="dv">100</span>))]</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>reduce_genes.by.immunologic.pathway<span class="ot">&lt;-</span>genes.by.immunologic.pathway[<span class="fu">unlist</span>(<span class="fu">lapply</span>(genes.by.immunologic.pathway,<span class="cf">function</span>(x) <span class="fu">length</span>(x)<span class="sc">&lt;</span><span class="dv">1000</span>))]</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) {</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>  reduce_genes.by.immunologic.pathway<span class="ot">&lt;-</span>scPagwas<span class="sc">::</span><span class="fu">reduce_pathway</span>(</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathway_seed=</span><span class="fu">names</span>(reduce_genes.by.immunologic.pathway)[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(reduce_genes.by.immunologic.pathway),<span class="dv">300</span>)],</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathway_list=</span>reduce_genes.by.immunologic.pathway,</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">remove_proporion=</span><span class="fl">0.1</span>)</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
<div id="gene-block-annotation" class="section level2">
<h2>4.Gene block annotation</h2>
<p>Gene block annotation in chromosome is need to prepare for scPagwas.
scPagwas can provide a block annotation data for protein-coding
genes.</p>
<p>the processed block annotation data are show here:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scPagwas))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(block_annotation))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">chrom</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="left">label</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">chr1</td>
<td align="right">11869</td>
<td align="right">14409</td>
<td align="left">DDX11L1</td>
</tr>
<tr class="even">
<td align="left">13</td>
<td align="left">chr1</td>
<td align="right">14404</td>
<td align="right">29570</td>
<td align="left">WASH7P</td>
</tr>
<tr class="odd">
<td align="left">26</td>
<td align="left">chr1</td>
<td align="right">17369</td>
<td align="right">17436</td>
<td align="left">MIR6859-1</td>
</tr>
<tr class="even">
<td align="left">29</td>
<td align="left">chr1</td>
<td align="right">29554</td>
<td align="right">31109</td>
<td align="left">MIR1302-2HG</td>
</tr>
<tr class="odd">
<td align="left">37</td>
<td align="left">chr1</td>
<td align="right">30366</td>
<td align="right">30503</td>
<td align="left">MIR1302-2</td>
</tr>
<tr class="even">
<td align="left">40</td>
<td align="left">chr1</td>
<td align="right">34554</td>
<td align="right">36081</td>
<td align="left">FAM138A</td>
</tr>
</tbody>
</table>
<p>Here is the procedure of obtaining the data:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;rtracklayer&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>gtf_df<span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="st">&quot;gencode.v34.annotation.gtf.gz&quot;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>gtf_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(gtf)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>gtf_df <span class="ot">&lt;-</span> gtf_df[,<span class="fu">c</span>(<span class="st">&quot;seqnames&quot;</span>,<span class="st">&quot;start&quot;</span>,<span class="st">&quot;end&quot;</span>,<span class="st">&quot;type&quot;</span>,<span class="st">&quot;gene_name&quot;</span>)]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>gtf_df <span class="ot">&lt;-</span> gtf_df[gtf_df<span class="sc">$</span>type<span class="sc">==</span><span class="st">&quot;gene&quot;</span>,]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>block_annotation<span class="ot">&lt;-</span>gtf_df[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>)]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(block_annotation)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>, <span class="st">&quot;start&quot;</span>,<span class="st">&quot;end&quot;</span>,<span class="st">&quot;label&quot;</span>)</span></code></pre></div>
</div>
<div id="ld-data" class="section level2">
<h2>5.LD data</h2>
<p>The 1,000 Genomes Project Phase 3 Panel was applied to calculate the
linkage disequilibrium (LD) among SNPs extracted from GWAS summary
statistics. the processed LD data are show here:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scPagwas))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(chrom_ld)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;list&quot;</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(chrom_ld[[<span class="st">&quot;chr1&quot;</span>]]))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">SNP_A</th>
<th align="left">SNP_B</th>
<th align="right">R</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">rs548419688</td>
<td align="left">rs568405545</td>
<td align="right">1.000000</td>
</tr>
<tr class="even">
<td align="left">rs575272151</td>
<td align="left">rs544419019</td>
<td align="right">1.000000</td>
</tr>
<tr class="odd">
<td align="left">rs62635286</td>
<td align="left">rs200579949</td>
<td align="right">1.000000</td>
</tr>
<tr class="even">
<td align="left">rs531646671</td>
<td align="left">rs541940975</td>
<td align="right">1.000000</td>
</tr>
<tr class="odd">
<td align="left">rs564003018</td>
<td align="left">rs374029747</td>
<td align="right">0.580208</td>
</tr>
<tr class="even">
<td align="left">rs565824523</td>
<td align="left">rs528394432</td>
<td align="right">1.000000</td>
</tr>
</tbody>
</table>
<p>We use <code>vcftools</code> and <code>plink</code> to deal with the
1,000 Genomes genotypes.</p>
<pre><code>./vcftools --vcf ./1000genomes_all_genotypes.vcf --plink-tped --out ./1000genomes_all_genotypes
./plink --tfile ./1000genomes_all_genotypes --recode --out ./1000genomes_all_genotypes
./plink --map ./1000genomes_all_genotypes.map --ped ./1000genomes_all_genotypes.ped --allow-no-sex --autosome --r2 --ld-window-kb 1000 --ld-window-r2 0.2 --out ./ld_1000genome</code></pre>
<p>R environment to print out the scPagwas-needed data.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>covid_ld<span class="ot">&lt;-</span><span class="fu">read.delim</span>(<span class="st">&quot;./ld_1000genome.ld&quot;</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#remove sex chrome</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>covid_ld<span class="ot">&lt;-</span>covid_ld[<span class="sc">!</span>(covid_ld<span class="sc">$</span> <span class="sc">%in%</span> <span class="dv">23</span>),]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(covid_ld)[<span class="dv">7</span>]<span class="ot">&lt;-</span><span class="st">&quot;R&quot;</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#print out the result in chrom number</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">unique</span>(covid_ld<span class="sc">$</span>CHR_A), <span class="cf">function</span>(i){</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  a<span class="ot">&lt;-</span><span class="fu">data.table</span>(covid_ld[covid_ld<span class="sc">$</span>CHR_A <span class="sc">==</span> i,])</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  file_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;./LD/&quot;</span>,i,<span class="st">&quot;.Rds&quot;</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(a, <span class="at">file =</span> file_name)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co">#integrate the data</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>chrom_ld<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>),<span class="cf">function</span>(chrom){</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  chrom_ld_file_path <span class="ot">&lt;-</span> <span class="fu">paste</span>(ld_folder, <span class="st">&#39;/&#39;</span>, chrom, <span class="st">&#39;.Rds&#39;</span>, <span class="at">sep =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a> ld_data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(chrom_ld_file_path)[(R<span class="sc">**</span><span class="dv">2</span> <span class="sc">&gt;</span> r2_threshold), .(SNP_A, SNP_B, R)]</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ld_data)</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Chunyu Deng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>
